src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +452,317 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+            {
+                // This expansion is not derived from a symbol
+                _state.ClearSymbolInformation();
+                return true;
+            }
+
+            if (!(SubjectBuffer.GetFeatureOnOffOption(CompletionOptions.EnableArgumentCompletionSnippets) ?? false))
+            {
+                // Argument completion snippets are not enabled
+                return false;
+            }
+
+            var symbols = ThreadingContext.JoinableTaskFactory.Run(() => GetSymbolsAsync(caretPosition: triggerSpan.End, cancellationToken));
+
+            var methodSymbols = symbols.OfType<IMethodSymbol>().ToImmutableArray();
+            if (methodSymbols.Any())
+            {
+                var methodName = dataBufferSpan.GetText();
+                var snippet = CreateSnippet(methodName, includeMethod: true, ImmutableArray<IParameterSymbol>.Empty, cancellationToken);
+
+                var doc = new DOMDocumentClass();
+                if (doc.loadXML(snippet.ToString(SaveOptions.OmitDuplicateNamespaces)))
+                {
+                    _state._symbols = methodSymbols;
+                    _state._symbol = null;
+
+                    if (expansion.InsertSpecificExpansion(doc, textSpan, this, LanguageServiceGuid, pszRelativePath: null, out _state._expansionSession) == VSConstants.S_OK)
+                    {
+                        Debug.Assert(_state._symbols == methodSymbols);
+                        Debug.Assert(_state._symbol == null);
+
+                        var componentModel = (IComponentModel)Package.GetGlobalService(typeof(SComponentModel));
+                        var controllerProvider = componentModel.DefaultExportProvider.GetExportedValue<ControllerProvider>();
+                        if (controllerProvider.GetController(TextView, SubjectBuffer) is { } controller)
+                        {
+                            // Register a handler for ModelUpdated. To avoid the possibility of more than one
+                            // handler being in the list, remove any current one before adding the new one. This
+                            // handler is only changed on the main thread, and the event is only invoked on the main
+                            // thread, so additional synchronization is not necessary to prevent lost events.
+                            controller.ModelUpdated -= OnModelUpdated;
+                            controller.ModelUpdated += OnModelUpdated;
+                        }
+
+                        // Trigger signature help after starting the snippet session
+                        //
+                        // TODO: Figure out why ISignatureHelpBroker.TriggerSignatureHelp doesn't work but this does.
+                        // https://github.com/dotnet/roslyn/issues/50036
+                        var editorCommandHandlerServiceFactory = componentModel.DefaultExportProvider.GetExportedValue<IEditorCommandHandlerServiceFactory>();
```

Comment:
why do we need to import this in this manner?  can't this just be normally MEF imported in the constructor? (or whatever constructor instantiates this client)?

Reply:
⏱ am switching to constructor injection